File: testing/backend/report.py
"""Generates a Markdown report for a completed run_all.py run."""
import datetime
from pathlib import Path

REPORTS_DIR = Path(__file__).resolve().parent / "reports"

_STATUS_ICON = {"PASS": "✅", "FAIL": "❌", "ERROR": "❌", "SKIP": "⏭️"}


def _escape_pipes(text):
    return text.replace("|", "\\|").replace("\n", " ")


def write_report(*, target, live_affecting, per_module, overall, elapsed, module_names):
    """Writes reports/backend_test.<timestamp>.md and returns its path."""
    REPORTS_DIR.mkdir(exist_ok=True)

    now = datetime.datetime.now(datetime.timezone.utc)
    stamp = now.strftime("%Y%m%d_%H%M%S")
    path = REPORTS_DIR / f"backend_test.{stamp}.md"

    overall_icon = "✅" if overall.failed == 0 else "❌"
    overall_verdict = "ALL PASSED" if overall.failed == 0 else f"{overall.failed} FAILURE(S)"

    lines = []
    lines.append("# Backend QA Suite — Run Report")
    lines.append("")
    lines.append("| | |")
    lines.append("|---|---|")
    lines.append(f"| **Generated** | {now.strftime('%Y-%m-%d %H:%M:%S UTC')} |")
    lines.append(f"| **Target** | `{target}` |")
    lines.append(f"| **Modules run** | {len(module_names)} |")
    lines.append(f"| **Live-affecting tests** | {'enabled' if live_affecting else 'disabled (default)'} |")
    lines.append(f"| **Duration** | {elapsed:.1f}s |")
    lines.append(f"| **Result** | {overall_icon} {overall_verdict} "
                 f"(pass={overall.passed}, fail={overall.failed}, skip={overall.skipped}) |")
    lines.append("")

    lines.append("## Summary")
    lines.append("")
    lines.append("| Module | Pass | Fail | Skip | Status |")
    lines.append("|---|---:|---:|---:|:---:|")
    for r in per_module:
        icon = "✅" if r.failed == 0 else "❌"
        lines.append(f"| {r.name} | {r.passed} | {r.failed} | {r.skipped} | {icon} |")
    lines.append(f"| **TOTAL** | **{overall.passed}** | **{overall.failed}** | **{overall.skipped}** | {overall_icon} |")
    lines.append("")

    if overall.failures:
        lines.append("## Failures")
        lines.append("")
        for label, message in overall.failures:
            lines.append(f"### {_escape_pipes(label)}")
            lines.append("")
            lines.append("```")
            lines.append(message)
            lines.append("```")
            lines.append("")

    lines.append("## Full results")
    lines.append("")
    for r in per_module:
        lines.append(f"### {r.name}")
        lines.append("")
        lines.append("| Case | Status | Time (ms) | Notes |")
        lines.append("|---|:---:|---:|---|")
        for c in r.cases:
            icon = _STATUS_ICON.get(c["status"], c["status"])
            note = _escape_pipes(c["message"]) if c["message"] else ""
            lines.append(
                f"| {_escape_pipes(c['description'])} | {icon} | {c['elapsed_ms']:.0f} | {note} |"
            )
        lines.append("")

    path.write_text("\n".join(lines) + "\n", encoding="utf-8")
    return path

File: frontend/report.mjs
#!/usr/bin/env node
// Turns a Vitest JSON-reporter file into a Markdown run report, mirroring the
// format and conventions of testing/backend/report.py so both suites read
// the same way. Usage: node report.mjs <path-to-vitest-json-output>
import { readFileSync, writeFileSync, mkdirSync } from 'node:fs'
import { dirname, join, relative } from 'node:path'
import { fileURLToPath } from 'node:url'

const HERE = dirname(fileURLToPath(import.meta.url))
const REPORTS_DIR = join(HERE, 'reports')
const STATUS_ICON = { passed: '✅', failed: '❌', pending: '⏭️', skipped: '⏭️', todo: '⏭️' }

function escapePipes(text) {
  return String(text).replace(/\|/g, '\\|').replace(/\r?\n/g, ' ')
}

function main() {
  const rawPath = process.argv[2]
  if (!rawPath) {
    console.error('Usage: node report.mjs <path-to-vitest-json-output>')
    return 1
  }
  const raw = JSON.parse(readFileSync(rawPath, 'utf-8'))
  mkdirSync(REPORTS_DIR, { recursive: true })

  const now = new Date()
  const stamp = now.toISOString().replace(/[-:]/g, '').replace('T', '_').slice(0, 15)
  const outPath = join(REPORTS_DIR, `frontend_test.${stamp}.md`)

  const files = raw.testResults ?? []
  const perFile = files.map((f) => {
    const assertions = f.assertionResults ?? []
    const passed = assertions.filter((a) => a.status === 'passed').length
    const failed = assertions.filter((a) => a.status === 'failed').length
    const skipped = assertions.length - passed - failed
    return { name: relative(HERE, f.name).replace(/\\/g, '/'), passed, failed, skipped, assertions }
  })

  const overallPassed = raw.numPassedTests ?? 0
  const overallFailed = raw.numFailedTests ?? 0
  const overallSkipped = (raw.numPendingTests ?? 0) + (raw.numTodoTests ?? 0)
  const latestEnd = files.reduce((max, f) => Math.max(max, f.endTime ?? 0), raw.startTime ?? 0)
  const durationSec = ((latestEnd - (raw.startTime ?? latestEnd)) / 1000).toFixed(1)
  const overallIcon = overallFailed === 0 ? '✅' : '❌'
  const overallVerdict = overallFailed === 0 ? 'ALL PASSED' : `${overallFailed} FAILURE(S)`

  const lines = []
  lines.push('# Frontend Test Suite — Run Report')
  lines.push('')
  lines.push('| | |')
  lines.push('|---|---|')
  lines.push(`| **Generated** | ${now.toISOString().replace('T', ' ').slice(0, 19)} UTC |`)
  lines.push('| **Suite** | Vitest unit/component tests (jsdom, mocked HTTP — no live backend, no test data written anywhere) |')
  lines.push(`| **Files run** | ${files.length} |`)
  lines.push(`| **Duration** | ${durationSec}s |`)
  lines.push(
    `| **Result** | ${overallIcon} ${overallVerdict} (pass=${overallPassed}, fail=${overallFailed}, skip=${overallSkipped}) |`
  )
  lines.push('')

  lines.push('## Summary')
  lines.push('')
  lines.push('| File | Pass | Fail | Skip | Status |')
  lines.push('|---|---:|---:|---:|:---:|')
  for (const f of perFile) {
    const icon = f.failed === 0 ? '✅' : '❌'
    lines.push(`| ${f.name} | ${f.passed} | ${f.failed} | ${f.skipped} | ${icon} |`)
  }
  lines.push(`| **TOTAL** | **${overallPassed}** | **${overallFailed}** | **${overallSkipped}** | ${overallIcon} |`)
  lines.push('')

  const failures = perFile.flatMap((f) =>
    f.assertions
      .filter((a) => a.status === 'failed')
      .map((a) => ({
        label: `${f.name} — ${a.fullName || a.title}`,
        message: (a.failureMessages || []).join('\n\n') || 'Failed',
      }))
  )
  if (failures.length) {
    lines.push('## Failures')
    lines.push('')
    for (const { label, message } of failures) {
      lines.push(`### ${escapePipes(label)}`)
      lines.push('')
      lines.push('```')
      lines.push(message)
      lines.push('```')
      lines.push('')
    }
  }

  lines.push('## Full results')
  lines.push('')
  for (const f of perFile) {
    lines.push(`### ${f.name}`)
    lines.push('')
    lines.push('| Case | Status | Time (ms) | Notes |')
    lines.push('|---|:---:|---:|---|')
    for (const a of f.assertions) {
      const icon = STATUS_ICON[a.status] || a.status
      const note = a.status === 'failed' ? escapePipes((a.failureMessages || [])[0] || '') : ''
      lines.push(`| ${escapePipes(a.fullName || a.title)} | ${icon} | ${(a.duration ?? 0).toFixed(0)} | ${note} |`)
    }
    lines.push('')
  }

  writeFileSync(outPath, lines.join('\n') + '\n', 'utf-8')
  console.log(`\nReport written to ${outPath}`)
  return overallFailed === 0 ? 0 : 1
}

process.exit(main())
